### PR TITLE
[Perf] flax_nnx: extend pre-flatten to compute_logits / embed_input_ids / combine_hidden_states / embed_multimodal / run_draft_model

### LINKED ISSUE #2607: [Performance] MODEL_IMPL_TYPE=flax_nnx is ~1.7× host-bound vs MODEL_IMPL_TYPE=vllm on decode-only
# `MODEL_IMPL_TYPE=flax_nnx` decode-only is ~1.7× host-bound vs `MODEL_IMPL_TYPE=vllm` due to per-dispatch `nnx.Variable` pytree flatten

## Summary

Running Gemma-4-31B-it / Qwen3-32B with `MODEL_IMPL_TYPE=flax_nnx` is **substantially slower per decode step than `MODEL_IMPL_TYPE=vllm`** in the decode-bound regime — for Gemma-4-31B-it, decode TPOT is 41.4 ms (flax_nnx) vs 28.1 ms (vllm-impl) at TP=2, batch=256.

Per-step profiling shows:

- On-TPU per-step time is actually **faster** for flax_nnx (~12.6 ms) than vllm-impl (~15.8 ms). The gap is entirely **host-side dispatch overhead**.
- ~90% of the 32 ms/step host gap lives inside the single `@jax.jit`'d `model_fn` dispatch.
- Root cause: `self.state` returned by `nnx.split(jit_model)` is an `nnx.State` whose leaves are wrapped in `nnx.Variable`. JAX walks this pytree on every jit dispatch and invokes Flax's `_variable_flatten` Python callback for every Variable — ~7,600 calls per step for Gemma-4-31B-it at 2.3 µs each.
- The `vllm` impl passes a plain dict-of-`jax.Array` (built via `jax_view(params_and_buffers)`), which JAX flattens through the C++ pytree fast path with no per-leaf Python callback.

Pre-flattening `state` once at init and passing the flat tuple to `model_fn` (reconstructing inside the jit via a closure'd treedef) closes 69% of the gap and brings decode TPOT to 26.2 ms — actually slightly faster than vllm-impl.

## Environment

- TPU host: `tpu7x-8` (8 chips, TP=2 used for most numbers)
- tpu-inference https://github.com/vllm-project/tpu-inference/commit/cae31c8b1d1eaa3228fcfc41ff533a290ca96c6c
- vLLM commit ea0e501bb18c12b80acc05ff8c7f013db515ba80

### Server commands

**Gemma-4-31B-it · `MODEL_IMPL_TYPE=flax_nnx`:**

```bash
USE_BATCHED_RPA_KERNEL=1 MODEL_IMPL_TYPE=flax_nnx vllm serve google/gemma-4-31B-it \
    --tensor-parallel-size 2 \
    --max-num-batched-tokens 4096 \
    --max-model-len 4096 \
    --no-enable-prefix-caching \
    --kv-cache-dtype=fp8 \
    --gpu-memory-utilization=0.9 \
    --async-scheduling \
    --limit-mm-per-prompt '{"image": 8, "video": 1, "audio": 0}' \
    --mm-processor-kwargs '{"size": {"longest_edge": 262144, "shortest_edge": 3136}}' \
    --block-size=256
```

**Gemma-4-31B-it · `MODEL_IMPL_TYPE=vllm`:** Adds two MM-encoder-pytree env vars and switches `MODEL_IMPL_TYPE`:

```bash
USE_BATCHED_RPA_KERNEL=1 \
JITTED_MM_MODULE_KEYS="model.vision_tower.encoder" \
REGISTER_MM_MODULE_CUSTOM_PYTREE_CLASSES="transformers.modeling_outputs.BaseModelOutputWithPast" \
MODEL_IMPL_TYPE=vllm vllm serve google/gemma-4-31B-it \
    <same flags as above>
```

**Qwen3-32B · `MODEL_IMPL_TYPE=flax_nnx`** (text-only; drop the MM flags):

```bash
USE_BATCHED_RPA_KERNEL=1 MODEL_IMPL_TYPE=flax_nnx vllm serve Qwen/Qwen3-32B \
    --tensor-parallel-size 2 \
    --max-num-batched-tokens 4096 \
    --max-model-len 4096 \
    --no-enable-prefix-caching \
    --kv-cache-dtype=fp8 \
    --gpu-memory-utilization=0.9 \
    --async-scheduling \
    --block-size=256
```

**Qwen3-32B · `MODEL_IMPL_TYPE=vllm`:** same flags as above with `MODEL_IMPL_TYPE=vllm`.

### Bench workloads

**Decode only** (1 in → 1024 out):

```bash
vllm bench serve --model <MODEL> --dataset-name random \
    --random-input-len 1 --random-output-len 1024 \
    --num-prompts 320 --ignore-eos --temperature=0
```

**Prefill only** (1024 in → 1 out):

```bash
vllm bench serve --model <MODEL> --dataset-name random \
    --random-input-len 1024 --random-output-len 1 \
    --num-prompts 320 --ignore-eos --temperature=0
```

**Image only** (Gemma-4 only — text-tokens=0, 512×512 image → 1 out):

```bash
vllm bench serve --backend openai-chat \
    --model google/gemma-4-31B-it --dataset-name random-mm \
    --num-prompts 320 --endpoint /v1/chat/completions \
    --random-mm-bucket-config '{(512, 512, 1): 1.0}' \
    --random-mm-limit-mm-per-prompt '{"image": 1}' \
    --random-mm-base-items-per-request 1 \
    --random-input-len 0 --random-output-len 1 --ignore-eos
```

## Baseline numbers

All decode/prefill rows are `Successful: 320/320`. (Qwen3-32B is text-only.)

| Workload | Metric | Gemma flax_nnx | Gemma vllm-impl | Qwen3 flax_nnx | Qwen3 vllm-impl |
|---|---|---:|---:|---:|---:|
| **Decode only** (320, 1→1024) | Output tok/s | 3922.81 | **6359** | 4913 | **7367** |
| | Mean TPOT (ms) | 41.39 | **28.06** | 33.45 | **23.78** |
| **Prefill only** (320, 1024→1) | Total tok/s | **21545** | 16699 | **23786** | 15215 |
| | Req/s | 21.02 | 16.29 | 23.21 | 14.84 |
| **Image only** (320, 0 txt + 512×512 → 1) | Total tok/s | **4140** | 745 | — | — |
| | Req/s | 15.22 | 2.74 | — | — |
| | Mean TTFT (ms) | 14911 | 89311 | — | — |

Headline: `flax_nnx` wins prefill / image / mixed workloads on Gemma. **`vllm` wins decode by a clear margin on both models** (~1.8× on Gemma, ~1.4× on Qwen3 by tok/s).

### Step 1: bracket the dispatch path

Added `jax.profiler.TraceAnnotation` spans around the substeps inside `tpu_inference/runner/tpu_runner.py:_execute_model`:

- `step.update_states`, `step.prepare_inputs`, `step.mm_encoder`, `step.mm_gather_embeds`, `step.get_input_embeds`, `step.extract_lora`, `step.model_fn`, `step.select_logits_indices`, `step.compute_logits`

Steady-state capture (256-req decode, TP=2, 12 s window, mean per call):

| span | flax_nnx | vllm-impl | Δ | A/B |
|---|---:|---:|---:|---:|
| `execute_model` | 80.66 ms | 48.38 ms | +32.28 | 1.67× |
| `step.model_fn` | 72.38 ms | 43.62 ms | **+28.75** | 1.66× |
| `step.compute_logits` | 4.48 ms | 0.67 ms | +3.81 | **6.70×** |
| `step.prepare_inputs` | 2.99 ms | 3.32 ms | −0.33 | 0.90 |
| `step.update_states` | 0.28 ms | 0.30 ms | ≈0 | 0.93 |
| (everything else) | <1 ms each, ≈ identical |

**~90% of the per-step gap is inside the single `model_fn` jit dispatch.** Both `model_fn` and `compute_logits_fn` take `self.state` as their first arg, which points the finger at how state is passed.

### Step 2: look at host-side perfetto events inside the jit wrapper

Top host events from the captures (mean per call):

| pjit / nnx internal | flax_nnx | vllm-impl |
|---|---:|---:|
| `PjitFunction(run_model)` / `PjitFunction(step_fun)` | 42.06 ms | 24.81 ms |
| `pjit.py:247 cache_miss` | 65.57 ms | 42.79 ms |
| `pjit.py:589 _infer_params` | 38.41 ms | (absorbed into `_run_python_pjit` 24.6 ms) |
| `flax/nnx/variablelib.py:2109 _variable_flatten` (count over 12 s window) | **30,362 calls / 70.42 ms total** | **0 calls** |

`_variable_flatten` is Flax's per-Variable pytree flatten callback. In flax_nnx it fires ~7,600 times *per decode step* at ~2.3 µs each → **~17.6 ms/step of pure Python pytree-flatten overhead**, contributing roughly half of the 32 ms gap. The rest is downstream pjit work (`_infer_params`) that scales linearly with the number of leaf descriptors.

### Step 3: confirm by pre-flattening once

Patched `tpu_inference/models/common/model_loader.py:get_flax_model` so that `state_leaves, _state_treedef = jax.tree_util.tree_flatten(state)` is computed once at init, and `run_model` accepts the flat tuple, with `nnx.merge(graphdef, tree_unflatten(_state_treedef, state_leaves))` happening inside the jit (so the rebuild is traced once, cached forever):

```python
graphdef, state = nnx.split(jit_model)
_state_leaves, _state_treedef = jax.tree_util.tree_flatten(state)
state_leaves = tuple(_state_leaves)

@jax.jit(
    out_shardings=(...),
    donate_argnums=1,       # kv_caches
    static_argnums=(6, 9, 10),
)
def run_model(state_leaves, *args):
    state = jax.tree_util.tree_unflatten(_state_treedef, state_leaves)
    model = nnx.merge(graphdef, state)
    return model(*args)
```

Plus a small plumbing change in `tpu_inference/models/common/interface.py` to expose `state_leaves` on `ModelInterface`, and `tpu_inference/runner/tpu_runner.py` + `tpu_inference/runner/compilation_manager.py` to pass `self.state_leaves` (instead of `self.state`) to the `model_fn` call sites. `run_compute_logits`, `embed_input_ids_fn`, `run_draft_model`, and eagle3's spec-decode paths are left untouched (still receive the original `nnx.State`).

#### Effect on the same capture

| | pre-fix flax_nnx | **post-fix flax_nnx** | vllm-impl |
|---|---:|---:|---:|
| `execute_model` mean | 80.66 ms | **58.42 ms** | 48.38 ms |
| `step.model_fn` mean | 72.38 ms | **50.08 ms** | 43.62 ms |
| `_variable_flatten` in window | 30,362 | **4,794** | 0 |
| `_variable_flatten` total ms | 70.4 | **7.1** | 0 |
| `_infer_params` mean per call | 38.4 ms | **20.9 ms** | 18.2 ms |
| `PjitFunction(run_model)` mean | 42.06 ms | **29.21 ms** | 24.81 ms |

Closed **69% of the per-step host gap** (22.3 of 32.3 ms). The residual ~10 ms = ~3.9 ms in `step.compute_logits` (same root cause, untouched) + ~6.5 ms in `step.model_fn` (the ~1,200 remaining per-step `_variable_flatten` callbacks + small `_infer_params`/`PjitFunction` overhead).

### PR DESCRIPTION
**Pull request by lk-chen**
FIXED #2607 

## Summary

Follow-up to #2615 — applies the same closure-captured-treedef +
flat-tuple pre-flatten pattern to the remaining state-receiving
dispatch fns on the flax_nnx path:

- `run_compute_logits`
- `jitted_embed_input_ids` / `run_embed_input_ids`
- `combine_hidden_states`
- `run_embed_multimodal`
- `run_draft_model`

All five now accept a `state_leaves` flat tuple of `jax.Array` (the
same form #2615 introduced for `model_fn`) and reconstruct
`nnx.State` via a closure-captured treedef inside the jit, so per-call
dispatch skips the `nnx.Variable` pytree walk for these paths too.

## Plumbing

- `ModelInterface` exposes `state_leaves` (populated by both
  `get_flax_model` and `get_vllm_model`).
- `TPUModelRunner` uses `model.state_leaves` directly; the LoRA-reload
  re-flatten path is preserved.
- `CompilationManager` precompile sites for `input_embeddings_merger`,
  `input_embeddings_merger_text_only`, and `compute_logits` pass
  `runner.state_leaves`.
- `MultiModalManager.embed_multimodal_fn` is fed `state_leaves`.
- `Eagle3Proposer` carries its own `self.state_leaves` (re-derived
  after the load-time `embed_tokens` mutation) and threads it through
  the jit-call boundaries (`_propose`, `_prepare_inputs`, the various
  `_*_draft_token_ids` helpers).

## Throughput delta — flax_nnx, vs #2615

Same setup as #2615 (TPU v7x, TP=2, fp8 KV cache, qwix fp8 weight
quant). Δ columns are vs the §5a numbers measured in #2615.

| Workload | Metric | Gemma-4-31B-it | Δ vs #2615 | Qwen3-32B | Δ vs #2615 |
|---|---|---:|---:|---:|---:|
| **MMMU-Pro** (1730 req, conc=32) | req/s | **1.56** | **+14%** (1.37) | — | — |
|  | total tok/s | **2188** | **+17%** (1866) | — | — |
|  | accuracy | 0.7104 | +0.4pt (0.7064) | — | — |
| **MMLU** (14042 req) | total tok/s | 21498 | −0.2% (21535) | 22633 | −0.1% (22666) |
|  | accuracy | 0.8458 | −0.1pt (0.8465) | 0.8127 | +0.1pt (0.812) |
| **Image only** (320 req, 512×512 → 1 out) | total tok/s | 5527 | +0.1% (5523) | — | — |
|  | mean TTFT (ms) | 9864 | −1.7% (10038) | — | — |
| **Prefill only** (320 req, 1024→1) | total tok/s | 22240 | −0.8% (22420) | 23855 | +0.0% (23847) |
| **Decode only** (320 req, 1→1024) | output tok/s | **6864** | **+18%** (5810) | **8643** | **+5%** (8230) |
|  | mean TPOT (ms) | **25.28** | **−12.5%** (28.90) | **20.54** | **−1.7%** (20.90) |

Headline: another **+18% / −12.5% TPOT** on Gemma flax_nnx decode on
top of #2615 (the residual ~3.9 ms that the issue draft predicted from
`step.compute_logits`). MMLU/prefill/image are within ±1% as expected —
the fix is host-side dispatch and doesn't touch the compute path.

Cumulative vs the pre-#2615 baseline: Gemma flax_nnx decode goes from
3248 tok/s / 50.10 ms to **6864 tok/s / 25.28 ms** (**+111%** throughput,
**−50%** TPOT).

## Test plan
- [x] Decode-only (1 in → 1024 out, batch 320) — Gemma-4-31B-it + Qwen3-32B, 320/320 success
- [x] Prefill-only (1024 in → 1 out) — Gemma + Qwen3, 320/320 success
- [x] MMLU (14042 req) — Gemma + Qwen3, accuracy unchanged
- [x] MMMU-Pro (1730 req, conc=32) — Gemma, accuracy 0.7104 (≈ #2615's 0.7064)
- [x] Image-only (0 txt + 512×512 → 1 out) — Gemma, 320/320 success

### CODE REVIEW

**Comment by lk-chen** [comment]
#2657
